### ISSUE TITLE
Startup time of istio-proxy causes some pods to crash

### ISSUE THREAD

**Issue by Stono** [istio/istio#3533]
Hey, 
I'm not sure if there's a huge amount you can do about this one, but I thought I'd highlight it anyway.

If you read through https://github.com/kubernetes/ingress-nginx/issues/1968 - you will see that the `nginx-ingress-controller` attempts to query the kubernetes api for the version almost immediately when it boots.

This works fine when we're not using istio-proxy, however when I do wrap the ingress controller with istio-proxy, it repeatedly crashed because it was trying to do so before istio-proxy was in a position to handle the intercepted traffic.

We worked around this by adding a retry/delay into the controller to give istio-proxy time to "boot", but it feels like something we shouldn't really need to be doing?

Thoughts?

**Comment by nmittler** [comment]
@PiotrSikora can you look into this one?

**Comment by Stono** [comment]
PS.  I'd imagine you could quite easily reproduce this by having a pod which has a startup script which just periodically curls `kubernetes.default.svc.cluster.local` until it works.

Remember; my setup is:
  - GKE 1.9.2
  - Istio 0.5.1 (auto injector)
  - RBAC
  - mTLS

**Comment by rshriram** [comment]
Is this only for ingress ? Does adding pod healthchecks work?

**Comment by Stono** [comment]
@rshriram I'm not sure how health checks would help here I'm afraid, this is explained in the linked issue but to explained what happens:

  - nginx-ingress-controller boots up
  - istio-proxy boots up (in parallel)
  - nginx-ingress-controller immediately tries to talk to kubernetes.default and fails (as iptables from the init-container route the traffic to istio-proxy, which fails to deal with the request), nginx-ingress-controller immediately exits with a non-zero status code.

To work around this, the maintainer of nginx-ingress-controller added a backoff retry to the kubernetes.default query, and in doing so we observed that the connection to kubernetes.default starts to work after about 30 seconds.

**Comment by rshriram** [comment]
See if you can add a health check to pilot pod for /v1/registration or some pseudo health check Api that makes pilot retrieve all data from kube. Then when an istio proxy talks to pilot, it would get the config data immediately as opposed to waiting on pilot to retrieve all endpoints from kube.

**Comment by Stono** [comment]
@rshriram that seems to resolve it, adding:

```
         livenessProbe:
           exec:
             command:
             - curl
             - -f
             - http://istio-pilot.istio-system:8080/v1/registration
```

As a `livenessProbe` to the `istio-proxy` container seems to fix is.

Can i just validate my understanding of what is going on here, does curling `/v1/registration` force pilot to reload its configuration from kube?

It that's the case, wouldn't it make sense to have envoy call this as soon as it boots, because it's a "new" things that pilot needs to pick up?

**Comment by costinm** [comment]
Interesting - if this helps we should definitely add it to the default
config.

But I would add it as a handler to the pilot agent, which also needs to
fetch
AZ.

Note that we made few changes to pilot to make it load faster - the
time for 1000 services went from >100 sec to ~3 in one test, but we should
still add readiness/liveness probes.

**Comment by Stono** [comment]
The AZ fetch currently doesn't work on mTLS clusters btw, so I guess that'd need to be resolved first too.

**Comment by nmittler** [comment]
@Stono In #4569 I added a new option to the startup script `-x` to allow you to exclude outbound IPs.  We now use this [in our nginx test](https://github.com/istio/istio/blob/master/tests/e2e/tests/simple/testdata/nginx/nginx-injected.yaml.tmpl#L448) to add an exclusion for kube master.  If you're doing manual injection, you can provide this value to istioctl via the `--excludeIPRanges` flag.

This will all be in `0.8.0`, so give it a try when it's released.  Cheers!

**Comment by Stono** [comment]
Hey @nmittler - looks good, thanks and the `-n` flag resolves the need for me to manually remove the iptables rule (yay).  Look forward to 0.8.0 to test it out.

**Comment by Stono** [comment]
@costinm do we want to track the `-x` before the `-i` ordering thing as part of this?

**Comment by sakshigoel12** [comment]
@Stono can you test this in the 0.8 release?

@nmittler what is pending to close this issue?

**Comment by Stono** [comment]
Hey @nmittler and @sakshigoel12. 

It's a slightly complicated one as improvements on both sides (istio and nginx) mean that the pods no longer crash (good).

The problem space is slightly different now, too.  Since 0.8, we've had to explicitly add a `DestinationRule` because the `mtlsExcludedServices` part doesn't appear to be working.  I'll raise a separate issue for that shortly.

The second problem is that `-x` is applied in the `istio init` container _after_ `-i`.  I'll raise another issue to explain that one too.

That way we can close this one off in favour of the more specific issues

**Comment by Stono** [comment]
OK i've created the two separate issues, there is maybe still a reason to keep this issue open however.  Ignoring the issues above, when a container starts that requires outbound comms almost immediately, there appears to be a window in which envoy is configuring that it is unable to do so, note this log from nginx:

```
❯ kn logs ingress-nginx-internal-controller-5ff98ccd58-tjnx2 --follow master
Booting nginx...(/nginx-ingress-controller --default-backend-service=ingress-nginx/default-http-backend --configmap=ingress-nginx/nginx-configuration-internal --tcp-services-configmap=ingress-nginx/tcp-services --udp-services-configmap=ingress-nginx/udp-services --annotations-prefix=nginx.ingress.kubernetes.io --publish-service=ingress-nginx/ingress-nginx-internal --ingress-class=nginx-internal --enable-ssl-chain-completion=false)
-------------------------------------------------------------------------------
NGINX Ingress controller
  Release:    0.15.0
  Build:      git-f3e9292c0
  Repository: https://github.com/aledbf/ingress-nginx
-------------------------------------------------------------------------------

I0605 07:50:39.561671       1 flags.go:162] Watching for ingress class: nginx-internal
W0605 07:50:39.561716       1 flags.go:165] only Ingress with class "nginx-internal" will be processed by this ingress controller
W0605 07:50:39.561869       1 flags.go:195] Check of SSL certificate chain is disabled (--enable-ssl-chain-completion=false)
nginx version: nginx/1.13.12
W0605 07:50:39.564301       1 client_config.go:533] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
I0605 07:50:39.564460       1 main.go:160] Creating API client for https://10.192.96.1:443
W0605 07:50:49.629902       1 main.go:201] it was required to retry 1 times before reaching the API server
```

Specifically the `W0605 07:50:49.629902       1 main.go:201] it was required to retry 1 times before reaching the API server`

I doubt there is a huge amount we can do about it, as the istio-proxy container would need to boot before the application container... so maybe we just accept that it's life and close this off :-)

**Comment by nmittler** [comment]
@Stono thanks for creating the issues ... we'll take a look.

Regarding [your comment above](https://github.com/istio/istio/issues/3533#issuecomment-394618842), I believe the issue is still nginx's ability to contact kube master immediately, correct? If so, wouldn't that be resolved by fixing the iptable issue, so that connections to kube master bypass envoy altogether?

**Comment by Stono** [comment]
Yes, it would 👌

**Comment by Stono** [comment]
Providing the health checks on envoy don't put it into a ready state until
it is fully configured.

Otherwise nginx will potentially start trying to route traffic before envoy
is ready

**Comment by nmittler** [comment]
@Stono Can health checking actually help here? IIUC, there isn't currently a way in k8s to make the startup of one container dependent on the successful startup (i.e. transition to "healthy") of another container (except for init containers, which won't work for the proxy).

**Comment by Stono** [comment]
Are you on the call in an hour?  Would be easier to explain in person

**Comment by nmittler** [comment]
Yeah I'll be there.

**Comment by Stono** [comment]
As discussed on our call, health checks for envoy to return ready once envoy has received initial configuration will solve this problem for me.

Perhaps related? https://github.com/istio/istio/issues/1425
